rfd/0034-clients-rotate-certs.md

```diff
@@ -0,0 +1,187 @@
+
+- Long lived certificates are less secure than short lived certificates.
+- The User CA may be rotated due to a security breach or a scheduled rotation. When this happens, all User certificates will lose access, after a grace period, and need to be manually refreshed.
+- Eventually, long lived certificates will expire after their TTL. Then they need to be manually refreshed, which puts a strain on administrators and could lead to downtime if neglected.
+
+## Details
+
+Cert agent will be a new standalone client with the responsibility of continuously renewing a set of certificates before they lose access. Before explaining how the agent will function, we need to create a new abstraction for certificates to enable the agent.
+
+### Cert Store
+
+The agent will interact with certificates through its Cert Store.
+
+A Cert Store can be used to:
+ - authenticate a client
+ - store renewed certificates
+ - alert when the store's certificates are altered
+ - provide its certificates expiration time
+
+```go
+type Store interface {
+  // extends client.Credentials to authenticate a client
+  client.Credentials
+  // store certs
+	Store(certs) error
+  // when current certs expire
+	Expires() time.Time
+  // used by the Auth server when signing new certificates
+  PublicKey() []byte
+  // signal that these certificates have been changed
+  Refresh() <- chan struct{}
+}
+```
+
+#### Formats
+
+The `Store` interface can be used to support a wide variety of certificate storage formats. 
+
+Initially, the following Cert Stores will be implemented:
+ - `PathStore` (uses direct cert paths, successor of `KeyPairCredentials`)
+ - `ProfileStore` (tsh profile)
+ - `IdentityFileStore`
+
+More stores can be added in the future to support:
+ - db/app/kube certs in tsh profile
+ - kubernetes secrets 
+
+#### Refresh
+
+The Cert Store will watch it's certificates for updates, whether it's a file change or something else. When this occurs, it will send a message on its `refresh` channel.
+
+Refresh only uses a single channel, so the `Store` can only be used by a single Cert Agent. Additional Cert Agents will be prevented from using a used `Store` object.
+
+### Cert Agent
+
+The Agent will:
+ - hold a Cert Store 
+ - hold a Client that is authenticated by the Cert Store
+ - watch for upcoming certificate expiration events (CA rotation or TTL expiration)
+ - renew the Cert Store's certificates before expiration events
+
+```go
+type Agent struct {
+  // client used to renew certificates and watch for expiration events
+  Client Client
+  // the agent's certificates store
+  CertStore Store
+  // TTL will define how long new certs generated through the Agent will 
+  // live for. Note that it will be limited on the server side by the
+  // user's max_session_ttl.
+  TTL time.Duration
+}
+```
+
+Note that the agent's Cert Store will also be used to authenticate the agent's own client, meaning the agent will act on behalf of the certificates' user.
```

Comment:
may be use multiple cert stores? then one agent can manage multiple certs

Reply:
The issue is that if the client uses different certs from the ones being renewed, there are some weird edge cases that I'd rather avoid confusing users with. It's definitely doable and I have some client refactors and other options I have in mind, but it may over complicate this RFD for now. I'll give it some more thought.